### PR TITLE
improve size column readability in archives tab

### PR DESCRIPTION
**Pull request by hariseldon78**
I find hard to read the sizes of the archives at a glance when they are all mixed units: usually the longer numbers represent smaller sizes, making it difficult to find the biggest archives. So I made a basic algorithm to select a good unit that can adequately represent all sizes, that makes it much easier to read. Also I made the column align to the right, which further improves readability.

### Related Issue
https://github.com/borgbase/vorta/issues/1595

### Motivation and Context
Readability

### How Has This Been Tested?
Tested on my archive list, which sizes goes from about 100 MB to 150 GB

### Screenshots (if appropriate):
image

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://vorta.borgbase.com/contributing/) guide.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

*I provide my contribution under the terms of the [license](./../../LICENSE.txt) of this repository and I affirm the [Developer Certificate of Origin][dco].*

[dco]: https://developercertificate.org/

### CODE REVIEW

File: src/vorta/utils.py [hunk 1]
```diff
@@ -240,7 +240,22 @@ def sort_sizes(size_list):
...
+        if metric
+        else (2**10, ['', 'Ki', 'Mi', 'Gi', 'Ti', 'Pi', 'Ei', 'Zi', 'Yi'])
+    )
+    min_size = min(s for s in size_list if isinstance(s, int))
+    # we tolerate a maximum precision of 1
+    n = 0
+    while abs(min_size / power**n > 0.1) and n + 1 < len(units):
+        n += 1
+    n = max(n - 1, 0)
+    return 1, units[n]
```
Comment (real-yfprojects) [inline src/vorta/utils.py:255]:
```suggestion
def find_best_size_formatting(sizes, metric=True, precision=1):
    power =  10**3 if metric else 2**10
    min_size = min((s for s in sizes if isinstance(s, int)), default=None)
    if min_size is None:
        return 1, 0
    n = math.ceil(math.log(min_size, power))
    return 1, n if min_size / power**n >= 0.1**precision else n-1
```

Comment (real-yfprojects) [reply src/vorta/utils.py:255]: This simplifies the code and fixes the error.

File: src/vorta/views/source_tab.py [hunk 2]
```diff
@@ -21,6 +21,10 @@ class SourceColumn:
 
 
 class SizeItem(QTableWidgetItem):
+    def __init__(self, s):
+        super().__init__(s)
+        self.setTextAlignment(int('0x82', 16))
```
Comment (real-yfprojects) [inline src/vorta/views/source_tab.py:26]: Please use the `Qt.Alignment` enum.

**Review by real-yfprojects — COMMENTED** [review]
Please add signature type hints and docstrings for `pretty_bytes` and `find_best_size_formatting`.

**Comment by real-yfprojects** [comment]
[Pre-commit](pre-commit.com) will format your code (including sorting imports) automatically.

File: src/vorta/utils.py (hunk 1 above)
Comment (hariseldon78) [reply src/vorta/utils.py:255]: thanks, I'll apply it asap. so I deduce it was crashing on an empty list.. what a rookie mistake!

Comment (real-yfprojects) [reply src/vorta/utils.py:255]: >  I deduce it was crashing on an empty list.. what a rookie mistake!

Yes, error message was `src/vorta/utils.py:249: ValueError: min() arg is an empty sequence`.

**Comment by hariseldon78** [comment]
unfortunately pre-commit is not working on my system, this is the error I receive: 
[pre-commit.log](https://github.com/borgbase/vorta/files/10774263/pre-commit.log)
the fixes I found talks about upgrading the isort package, but the version is not pinned in the requirements file, so it should already be at the highest version compatible with the other packages...

**Comment by hariseldon78** [comment]
now that I have pre-commit running, if I run it with --all-files it finds a bunch of other little linting problems; should I commit those? how can I see the github builder error logs?

**Comment by real-yfprojects** [comment]
> [Edit] ok this problem is related to https://github.com/borgbase/vorta/pull/1601, so i've been able to install and run it by doing the `pre-commit autoupdate`

Rebasing your branch onto master will also do the job.

File: src/vorta/views/archive_tab.py [hunk 3]
```diff
@@ -243,12 +250,16 @@ def populate_from_profile(self):
 
             sorting = self.archiveTable.isSortingEnabled()
             self.archiveTable.setSortingEnabled(False)
+            precision = 1  # could be a setting
```
Comment (real-yfprojects) [inline src/vorta/views/archive_tab.py:253]: I think, a (global) constant would fit well for this.

File: src/vorta/utils.py [hunk 4]
```diff
@@ -249,10 +268,14 @@ def pretty_bytes(size, metric=True, sign=False, precision=1):
...
     )
-    n = 0
-    while abs(round(size, precision)) >= power and n + 1 < len(units):
-        size /= power
-        n += 1
+    if fixed_unit is None:
+        n = 0
+        while abs(round(size, precision)) >= power and n + 1 < len(units):
+            size /= power
+            n += 1
```
Comment (real-yfprojects) [inline src/vorta/utils.py:275]: Can this be convertes to a one/twoliner as well?

File: src/vorta/utils.py [hunk 5]
```diff
@@ -249,10 +268,14 @@ def pretty_bytes(size, metric=True, sign=False, precision=1):
...
+    else:
+        n = fixed_unit
+        size /= power**n
     try:
         unit = units[n]
```
Comment (real-yfprojects) [inline src/vorta/utils.py:280]: What if index out of range? See comment above.

File: src/vorta/utils.py [hunk 6]
```diff
@@ -240,7 +241,25 @@ def sort_sizes(size_list):
...
+    """
+    Selects the index of the biggest unit (see the lists in the pretty_bytes function) capable of
+    representing the smallest size in the sizes iterable.
+    """
+    power = 10**3 if metric else 2**10
+    min_size = min((s for s in sizes if isinstance(s, int)), default=None)
+    if min_size is None:
+        return 0
+    n = math.floor(math.log(min_size * 10**precision, power))
+    return n
```
Comment (real-yfprojects) [inline src/vorta/utils.py:254]: n might be of metric range, i think

File: src/vorta/utils.py [hunk 7]
```diff
@@ -240,7 +241,25 @@ def sort_sizes(size_list):
...
+def find_best_size_unit(sizes: Iterable[int], metric: bool = True, precision: int = 1) -> int:
+    """
+    Selects the index of the biggest unit (see the lists in the pretty_bytes function) capable of
+    representing the smallest size in the sizes iterable.
+    """
+    power = 10**3 if metric else 2**10
+    min_size = min((s for s in sizes if isinstance(s, int)), default=None)
+    if min_size is None:
+        return 0
+    n = math.floor(math.log(min_size * 10**precision, power))
```
Comment (real-yfprojects) [inline src/vorta/utils.py:253]: Neat!

**Review by real-yfprojects — COMMENTED** [review]
Great, you added test cases and cleverly improved my code!

File: src/vorta/utils.py (hunk 5 above)
Comment (hariseldon78) [reply src/vorta/utils.py:280]: right, better clamp it

File: src/vorta/utils.py (hunk 6 above)
Comment (hariseldon78) [reply src/vorta/utils.py:254]: not sure what you mean... those lines should work both for metric and power of 2, do you think I'm missing some test cases?

Comment (real-yfprojects) [reply src/vorta/utils.py:254]: Sorry left out some words, updated the comment.

Comment (hariseldon78) [reply src/vorta/utils.py:270]: ok, yes i resolved by clamping in the pretty_bytes function

**Review by real-yfprojects — COMMENTED** [review]
Can you add a test for a very large size like $10^{30}$?

**Comment by hariseldon78** [comment]
> Can you add a test for a very large size like 1030?

it was failing actually :)

File: src/vorta/utils.py [hunk 8]
```diff
@@ -240,7 +241,41 @@ def sort_sizes(size_list):
...
+    representing the passed size.
+    """
+    if not isinstance(size, int) or size == 0:  # this will also take care of the None case
+        return 0
+    power = 10**3 if metric else 2**10
+    n = math.floor(math.log(size * 10**precision, power))
+    return n
+
+
+def pretty_bytes(size: int, metric: bool = True, sign: bool = False, precision: int = 1, fixed_unit: int = None) -> str:
```
Comment (real-yfprojects) [inline src/vorta/utils.py:273]:
```suggestion
def pretty_bytes(size: int, metric: bool = True, sign: bool = False, precision: int = 1, fixed_unit: Optional[int] = None) -> str:
```

File: src/vorta/views/archive_tab.py [hunk 9]
```diff
@@ -44,6 +51,8 @@
 
 logger = logging.getLogger(__name__)
 
+PRECISION = 1
```
Comment (real-yfprojects) [inline src/vorta/views/archive_tab.py:54]: Please add a docstring for the constant.

Comment (hariseldon78) [reply src/vorta/views/archive_tab.py]: Not sure how to do that, only way I found is the Annotated typing, but that needs python 3.9 at least (https://stackoverflow.com/a/8820636/527935). So instead I renamed the constant to `SIZE_DECIMAL_DIGITS` which seems more clear

Comment (real-yfprojects) [reply src/vorta/views/archive_tab.py]: Sphinx allows documenting python variables. I always use [their `#:`-format](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoattribute).

Comment (hariseldon78) [reply src/vorta/views/archive_tab.py]: ok. I still find the new name self explanatory enough, don't you think?

Comment (real-yfprojects) [reply src/vorta/views/archive_tab.py]: Method, function and class names should be self-explanatory too. However docstrings are mandatory still.

Comment (hariseldon78) [reply src/vorta/views/archive_tab.py]: ok done. If they are mandatory then please update the Contributing guide.

Comment (real-yfprojects) [reply src/vorta/views/archive_tab.py]: > If they are mandatory then please update the Contributing guide.

Quoting the docs:

> Write docstrings at least for methods and functions. Docstrings for classes, attributes and global variables are very welcome too.

So docstrings are mandatory for methods and functions, although you are right that I didn't write the same about other docstrings in the guide. In these cases I will request them as applicable.

File: src/vorta/utils.py [hunk 10]
```diff
@@ -240,7 +241,43 @@ def sort_sizes(size_list):
...
+def find_best_unit_for_sizes(sizes: Iterable[int], metric: bool = True, precision: int = 1) -> int:
+    """
+    Selects the index of the biggest unit (see the lists in the pretty_bytes function) capable of
+    representing the smallest size in the sizes iterable.
+    """
+    min_size = min((s for s in sizes if isinstance(s, int)), default=None)
+    return find_best_unit_for_size(min_size, metric=metric, precision=precision)
+
+
+def find_best_unit_for_size(size: int, metric: bool = True, precision: int = 1) -> int:
```
Comment (real-yfprojects) [inline src/vorta/utils.py:261]:
```suggestion
def find_best_unit_for_size(size: Optional[int], metric: bool = True, precision: int = 1) -> int:
```

File: src/vorta/utils.py [hunk 11]
```diff
@@ -240,7 +241,43 @@ def sort_sizes(size_list):
     return final_list
 
 
-def pretty_bytes(size, metric=True, sign=False, precision=1):
+T = TypeVar("T")
+
+
+def clamp(n: T, min_: T, max_: T) -> T:
+    """Restrict the number n inside a range"""
+    return min(max_, max(n, min_))
```
Comment (real-yfprojects) [inline src/vorta/utils.py:249]:
```suggestion
    return min(max_, max(n, min_))  # type: ignore
```

**Review by real-yfprojects — COMMENTED** [review]
Only some minor typing problems which make them even more minor.

File: src/vorta/utils.py (hunk 11 above)
Comment (hariseldon78) [reply src/vorta/utils.py:249]: ok I activated mypy in my emacs config, didn't know about that checker! I added the `int, float` specification to the typevar and now it doesn't complain

**Review by real-yfprojects — APPROVED** [review]

**Comment by m3nu** [comment]
Thanks all! Give me until this evening to run this locally.

**Comment by m3nu** [comment]
Works as advertised. Interesting addition. Thanks for your work and patience on this!

**Review by m3nu — APPROVED** [review]

**Pull request merged**